contracts/oracle/CollateralizationOracle.sol

```diff
@@ -0,0 +1,287 @@
+    /// @return _globalValid the current oracle validity status (false if any
+    ///         of the oracles for tokens held in the PCV are invalid, or if
+    ///         this contract is paused).
+    function read() public override view returns (Decimal.D256 memory, bool) {
+        // The total amount of FEI controlled (owned) by the protocol
+        uint256 _protocolControlledFei = 0;
+        // The total USD value of assets held in the PCV
+        uint256 _protocolControlledValue = 0;
+        // false if any of the token oracles are invalid, or if paused
+        bool _globalValid = !paused();
+
+        // For each token...
+        for (uint256 i = 0; i < tokensInPcv.length; i++) {
+            address _token = tokensInPcv[i];
+            (Decimal.D256 memory _price, bool _valid) = IOracle(token2oracle[_token]).read();
+            _globalValid = _globalValid && _valid;
+
+            // For each deposit...
+            for (uint256 j = 0; j < token2deposits[_token].length; j++) {
+                IPCVDepositV2 _deposit = IPCVDepositV2(token2deposits[_token][j]);
+
+                // Increment the total protocol controlled value by the USD value of
+                // the asset held in this PCVDeposit
+                _protocolControlledValue += _price.mul(_deposit.resistantBalance()).asUint256();
+
+                // Increment the total protocol controlled FEI by the amount of FEI
+                // held by this PCVDeposit
+                _protocolControlledFei += _deposit.resistantProtocolOwnedFei();
+            }
+        }
+
+        // The circulating FEI is defined as the total FEI supply, minus the
+        // protocol-owned FEI managed by the various deposits.
+        uint256 _userCirculatingFei = fei().totalSupply() - _protocolControlledFei;
+
+        // The protocol collateralization ratio is defined as the total USD
+        // value of assets held in the PCV, minus the circulating FEI.
+        Decimal.D256 memory _collateralRatio = Decimal.ratio(_protocolControlledValue, _userCirculatingFei);
+
+        // return values
+        return (_collateralRatio, _globalValid);
+    }
+
+    // ----------- ICollateralizationOracle override methods -----------
+
+    /// @notice returns true if the protocol is overcollateralized. Overcollateralization
+    ///         is defined as the protocol having more assets in its PCV (Protocol
+    ///         Controlled Value) than the circulating (user-owned) FEI.
+    function isOvercollateralized() external view override whenNotPaused returns (bool) {
+        return pcvEquityValue() > 0;
+    }
+
+    /// @notice returns the amount of circulating (user-owned) FEI. User Circulating
+    ///         FEI is defined as the total supply of FEI minus the FEI held in the
+    ///         various protocol's contracts.
+    function userCirculatingFei() external view override whenNotPaused returns (uint256) {
```

Comment:
I think it would be best to combine all of these functions into a single one that returns userCirculatingFei, pcvValue and pcvEquity to save gas and simplify the contract. The `read` function can just call that one and do the final logic step